engine/src/collisions/narrow_phase/plugin.cpp

```diff
@@ -113,78 +198,229 @@ void cubos::engine::narrowPhaseCollisionsPlugin(Cubos& cubos)
+                                                      .position = {0.0F, 0.0F, 0.0F},
+                                                      .normal = intersectionInfo.normal});
+                        }
+                    }
+                }
+            }
+        });
 
-                for (const glm::vec3& point : polygon2)
+    cubos.system("collision manifolds voxel-box")
+        .tagged(collisionsManifoldTag)
+        .after(collisionsNarrowTag)
+        .call([](Commands cmds, Query<Entity, const LocalToWorld&, const BoxCollisionShape&, CollidingWith&, Entity,
+                                      const LocalToWorld&, const VoxelCollisionShape&>
+                                    query) {
+            for (auto [ent1, localToWorld1, boxShape, collidingWith, ent2, localToWorld2, voxelShape] : query)
+            {
+                // If penetration not bigger than 0 continue
+                if (collidingWith.penetration < 0)
                 {
-                    // Compute distance to reference plane
-                    glm::vec3 pointDiff = point - cubos::core::geom::getClosestPointPolygon(point, polygon1);
-                    float contactPenetration = -glm::dot(pointDiff, collidingWith.normal); // is this positive
+                    continue;
+                }
 
-                    // Set Contact data
-                    glm::vec3 globalOn1 = point;                                             // world coordinates
-                    glm::vec3 globalOn2 = point + collidingWith.normal * contactPenetration; // world coordinates
+                std::vector<ContactPointData> points;
+                // Make sure that shape1 corresponds to the entity refered to in collidingWith
+                if (ent1 == collidingWith.entity)
+                {
+                    const cubos::core::geom::Box* matchedShape1 = &boxShape.box;
+                    const LocalToWorld* matchedLocalToWorld1 = &localToWorld1;
+                    for (auto box : voxelShape.getBoxes())
+                    {
+                        const cubos::core::geom::Box* matchedShape2 = &box.box;
+                        LocalToWorld matchedLocalToWorld2 = localToWorld2;
+
+                        // Create a translation matrix for the shift
+                        glm::mat4 shiftMatrix = glm::translate(glm::mat4(1.0F), -box.shift);
+                        matchedLocalToWorld2.mat *= shiftMatrix;
 
-                    // If we flipped incident and reference planes, we will
-                    // need to flip it back before sending it to the manifold.
-                    if (flipped)
+                        auto newPoints = computePoints(matchedShape1, matchedLocalToWorld1, matchedShape2,
+                                                       &matchedLocalToWorld2, collidingWith);
+                        points.insert(points.end(), newPoints.begin(), newPoints.end());
```

Comment:
Correct me if I'm wrong, here you are going through all boxes in the voxel shape and computing the manifold with the box, right? However, the collidingWith was already performed between each of these, which means that we should know exactly which boxes withing the voxel shape are colliding with this one. I think that we could maybe keep like an index or whatever is best of the specific box in the voxel shape that we collided with and only do computePoints between that and the box. 

Extra, but for more context: The computation of the manifold assumes the shapes are colliding, the only sanity check it performs is with penetration in the direction of the normal, and we are computing it between boxes that may not be colliding, I think this can create incorrect forces in the solver in some occasions.

Reply:
Disregard my comment here, I think it's something still worth thinking about but we have to change other things, before worrying about this.